support/refc/_datatypes.h

```diff
@@ -50,15 +45,31 @@ typedef struct {
   // followed by type-specific payload.
 } Value;
 
-typedef struct {
-  Value_header header;
-  uint8_t ui8;
-} Value_Bits8;
-
-typedef struct {
-  Value_header header;
-  uint16_t ui16;
-} Value_Bits16;
+/* expcted at least 4bytes for Value_header alignment. */
+#define idris2_vp_is_unboxed(p) ((uintptr_t)(p)&3)
+
+#define idris2_vp_int_shift                                                    \
+  ((sizeof(uintptr_t) >= 8 && sizeof(Value *) >= 8) ? 32 : 16)
+
+#define idris2_vp_to_Bits64(p) (((Value_Bits64 *)(p))->ui64)
+#define idris2_vp_to_Bits32(p)                                                 \
+  ((idris2_vp_int_shift == 16)                                                 \
+       ? (((Value_Bits32 *)(p))->ui32)                                         \
+       : ((uint32_t)((uintptr_t)(p) >> idris2_vp_int_shift)))
```

Comment:
This and similar checks seems to be runtime checks. Can't we turn these to compile-time ones, say with

```c
#if INTPTR_WIDTH >= 32
#define idris2_vp_to_Bits32(p) /* direct unboxed use */
#else
#define idris2_vp_to_Bits32(p) (((Value_Bits32 *)(p))->ui32)
#endif
```

Or, do you rely on the compiler's optimisation on the constant `if` conditions?

Reply:
> Or, do you rely on the compiler's optimisation on the constant if conditions?

Yes, An optimizer of a modern C compiler is awesome. However INTPTR_WIDTH  or UINTPTR_WIDTH that I forgot one sounds better to me, I'll try it.  thanks!

Reply:
After a little experimentation, I found that UINTPTR_WIDTH is an optional feature of C23. And some of my target compilers did not yet support it.
So I prepared a fallback just in case.